File: frontend/src/lib/locations/glyphs.ts
/**
 * The pictogram a container is drawn with — the cheap half of "what does this
 * look like", the other half being the photo `ContainerPhoto` renders (see that
 * component's docstring for the split and why the dense tree never loads one).
 *
 * Deliberately the *sibling* of `lib/locations/views.ts`, not a copy of it: this
 * module only ever answers "what character represents this glyph name", because
 * `effective_glyph` — the instance override, else the container type's, else
 * `null` — is already resolved server-side (`app.services.glyphs`) for exactly
 * the reason `effective_child_view` is: the fallback would otherwise need
 * `container_types` fetched and joined again on this side.
 *
 * `null`/unknown both mean "no glyph" here, and that is a real answer rather
 * than a loading state: unlike `child_view`, there is no derivation underneath
 * a glyph, so a container with neither its own nor its type's is a container
 * nobody has picked a picture for yet.
 */

import type { ContainerGlyph } from "../api/client";

/** A single character per glyph — cheap enough to sit inside a slot label's
 * own cell in a dense grid without a second network fetch. */
export const GLYPH_SYMBOLS: Readonly<Record<ContainerGlyph, string>> = {
  box: "📦",
  bin: "🗑",
  drawer: "🗄",
  cabinet: "🚪",
  shelf: "📚",
  tray: "🍱",
  bag: "👜",
  reel: "🧵",
  room: "🏠",
  rack: "🎛",
};

export const GLYPH_LABELS: Readonly<Record<ContainerGlyph, string>> = {
  box: "Box",
  bin: "Bin",
  drawer: "Drawer",
  cabinet: "Cabinet",
  shelf: "Shelf",
  tray: "Tray",
  bag: "Bag",
  reel: "Reel",
  room: "Room",
  rack: "Rack",
};

/**
 * The character to render for a resolved glyph name, or `null` for "draw the
 * neutral placeholder instead" — which covers both "no glyph was ever chosen"
 * and "a newer build named one this bundle has never heard of". The second
 * case is the other half of the no-`CHECK` promise: `container_types.glyph`
 * and `locations.glyph` carry no constraint, so a row can legally hold a name
 * this bundle predates, and the honest response is the same placeholder an
 * unset glyph gets — never a crash.
 */
export function glyphSymbol(glyph: string | null): string | null {
  if (glyph === null || !(glyph in GLYPH_SYMBOLS)) {
    return null;
  }
  return GLYPH_SYMBOLS[glyph as ContainerGlyph];
}

export function glyphLabel(glyph: string | null): string | null {
  if (glyph === null || !(glyph in GLYPH_LABELS)) {
    return null;
  }
  return GLYPH_LABELS[glyph as ContainerGlyph];
}

/** Every glyph a picker can offer, in the fixed display order above. */
export const ALL_GLYPHS: readonly ContainerGlyph[] = Object.keys(
  GLYPH_LABELS,
) as ContainerGlyph[];

File: backend/app/services/glyphs.py
"""Which pictogram a container is drawn with.

Two rungs, not three — the deliberate contrast with `app.services.views`, which
this module is otherwise a sibling of:

    instance override  →  locations.glyph
    type default       →  container_types.glyph

There is no derived rung. `resolve_child_view` can fall back to reading a type's
declared geometry because a 42 mm pitch really does imply "a tray seen from
above"; nothing about a type's geometry implies what it *looks like*, so a
"derived glyph" would be a guess wearing the shape of a fact. `None` all the way
down is therefore a real, terminal answer — "no glyph chosen" — and the caller
(the frontend's dense recursive map, `LocationRead`, `ContainerTypeRead`) renders
a neutral placeholder for it rather than inventing one.

**Returns `str | None`, not `ContainerGlyph | None`.** Exactly the promise
`resolve_child_view` makes for the same reason: a glyph name written by a newer
build passes through untouched rather than raising in `ContainerGlyph(...)`, and
the frontend's own fallback map (`frontend/src/lib/locations/glyphs.ts`) treats an
unrecognised name the same as `None` — a placeholder, never a crash.
"""

from __future__ import annotations

from collections.abc import Iterable

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.storage import ContainerType, Location


def resolve_glyph(location: Location | None, container_type: ContainerType | None) -> str | None:
    """The effective glyph for one container: its own override, else its
    type's, else `None`."""
    if location is not None and location.glyph is not None:
        return location.glyph
    if container_type is not None:
        return container_type.glyph
    return None


def resolve_glyphs(session: Session, locations: Iterable[Location]) -> dict[int, str | None]:
    """`resolve_glyph` for a whole tree, in one extra query.

    Same shape as `app.services.views.resolve_child_views`, and for the same
    reason: `GET /api/locations/tree` returns every node in one response, so
    resolving per node would be one `container_types` lookup per location.
    """
    rows = list(locations)
    type_ids = {row.container_type_id for row in rows if row.container_type_id is not None}
    types: dict[int, ContainerType] = {}
    if type_ids:
        types = {
            container_type.id: container_type
            for container_type in session.execute(
                select(ContainerType).where(ContainerType.id.in_(type_ids))
            ).scalars()
        }
    return {
        row.id: resolve_glyph(
            row, types.get(row.container_type_id) if row.container_type_id is not None else None
        )
        for row in rows
    }
